src/NATS.Client.Services/NatsSvcService.cs

```diff
@@ -0,0 +1,329 @@
+    private readonly Task _taskMsgLoop;
+    private readonly List<SvcListener> _svcListeners = new();
+    private readonly ConcurrentDictionary<string, INatsSvcEndPoint> _endPoints = new();
+    private readonly string _started;
+    private readonly CancellationTokenSource _cts;
+
+    /// <summary>
+    /// Creates a new instance of <see cref="NatsSvcService"/>.
+    /// </summary>
+    /// <param name="nats">NATS connection.</param>
+    /// <param name="config">Service configuration.</param>
+    /// <param name="cancellationToken">A <see cref="CancellationToken"/> used to cancel the service creation requests.</param>
+    public NatsSvcService(NatsConnection nats, NatsSvcConfig config, CancellationToken cancellationToken)
+    {
+        _logger = nats.Opts.LoggerFactory.CreateLogger<NatsSvcService>();
+        _id = NuidWriter.NewNuid();
+        _nats = nats;
+        _config = config;
+        _cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+        _cancellationToken = _cts.Token;
+        _channel = Channel.CreateBounded<SvcMsg>(32);
+        _taskMsgLoop = Task.Run(MsgLoop);
+        _started = DateTimeOffset.UtcNow.ToString("yyyy-MM-ddTHH:mm:ss.fffffffZ");
+    }
+
+    /// <summary>
+    /// Stop the service.
+    /// </summary>
+    /// <param name="cancellationToken">A <see cref="CancellationToken"/> used to cancel the stop operation.</param>
+    /// <returns>A <seealso cref="ValueTask"/> representing the asynchronous operation.</returns>
+    public async ValueTask StopAsync(CancellationToken cancellationToken = default)
+    {
+        foreach (var listener in _svcListeners)
+        {
+            await listener.DisposeAsync();
+        }
+
+        // Drain buffers
+        await _nats.PingAsync(cancellationToken);
+
+        foreach (var ep in _endPoints.Values)
+        {
+            await ep.DisposeAsync();
+        }
+
+        _channel.Writer.TryComplete();
+
+        _cts.Cancel();
+
+        await _taskMsgLoop;
+    }
+
+    /// <summary>
+    /// Adds a new endpoint.
+    /// </summary>
+    /// <param name="handler">Callback for handling incoming messages.</param>
+    /// <param name="name">Optional endpoint name.</param>
+    /// <param name="subject">Optional endpoint subject.</param>
+    /// <param name="metadata">Optional endpoint metadata.</param>
+    /// <param name="cancellationToken">A <see cref="CancellationToken"/> used to stop the endpoint.</param>
+    /// <typeparam name="T">Serialization type for messages received.</typeparam>
+    /// <returns>A <seealso cref="ValueTask"/> representing the asynchronous operation.</returns>
+    /// <remarks>
+    /// One of name or subject must be specified.
+    /// </remarks>
+    public ValueTask AddEndPointAsync<T>(Func<NatsSvcMsg<T>, ValueTask> handler, string? name = default, string? subject = default, IDictionary<string, string>? metadata = default, CancellationToken cancellationToken = default) =>
```

Comment:
Nit - I am a fan of "endpoint" as one word so the capitalization should probably be `Endpoint` wherever this appears

Yes... `System.Net.EndPoint` does "end point"...  But `Microsoft.AspNetCore.Http.Endpoint` does "endpoint" :smile:

Reply:
Hehe yes I thought about that, then checked IPEndPoint 😆 I agree. 'endpoint' is a word so..